drain3/template_miner.py

```diff
@@ -170,24 +171,93 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
+            if exact_matching:
+                mask_patterns.setdefault(mask, [])
+                mask_patterns[mask].append((instruction.regex_pattern, list(instruction.regex.groupindex.keys())))
+            else:
+                mask_patterns[mask] = [simple_parameter_spec]
+
+        # Add a pattern for masks introduced by the Drain parser.
+        mask_patterns.setdefault("*", [])
+        mask_patterns["*"].append(simple_parameter_spec)
+
+        # Prepare for mask-substitution.
         escaped_prefix = re.escape(self.config.mask_prefix)
         escaped_suffix = re.escape(self.config.mask_suffix)
-        template_regex = re.sub(escaped_prefix + r".+?" + escaped_suffix, self.drain.param_str, log_template)
-        if self.drain.param_str not in template_regex:
-            return []
-        template_regex = re.escape(template_regex)
-        template_regex = re.sub(r'\\ +', r'\\s+', template_regex)
-        template_regex = "^" + template_regex.replace(escaped_prefix + r"\*" + escaped_suffix, "(.*?)") + "$"
+        template_regex = re.escape(log_template)
+        parameter_groups = dict()
+        named_groups = set()
+        def unique_name(original):
+            while original in named_groups:
+                original += str(random.randint(0, 100000))
+            named_groups.add(original)
+            return original
+
+        # Mask-substitution: For every mask in the template, replace it with
+        # a named group of all possible masking-patterns it could represent (in order).
+        for mask, patterns in mask_patterns.items():
+            def create_parameter_group(match):
+                modified_patterns = []
+                for pattern, groups in patterns:
+                    # Any pattern may named groups that have already been used.
+                    # Therefore the affected groups are given a new unique name.
+                    for group_name in groups:
+                        if group_name not in named_groups:
+                            continue
+                        new_name = unique_name(group_name)
+                        pattern = re.sub(r"\(\?P={}\)".format(re.escape(group_name)), lambda x: "(?P={})".format(new_name), pattern)
+                        pattern = re.sub(r"\(\?P<{}>".format(re.escape(group_name)), lambda x: "(?P<{}>".format(new_name), pattern)
+                    named_groups.update(groups)
+                    modified_patterns.append(pattern)
+                parameter_group = unique_name("parameter")
+                parameter_groups[parameter_group] = mask
+                return "(?P<{}>".format(parameter_group) + "|".join(modified_patterns) + ")"
+            template_regex = re.sub(re.escape(escaped_prefix) + re.escape(re.escape(mask)) + re.escape(escaped_suffix),
+                                    create_parameter_group, template_regex)
+
+        template_regex = re.sub(r"\\ ", r"\\s+", template_regex)
+        template_regex = "^" + template_regex + "$"
```

Comment:
Adding a cache indexed by template for `template_regex` might speed up the function by a more considerable amount, since the process of creating this pattern does not need to be run each time the same template is applied.